Update CONTRIBUTING.md (#1365)
cp-algorithms/cp-algorithms: CONTRIBUTING.md

---
search:
  exclude: true

> ---

Author: Yes, keeping this content in the search index is important because it helps contributors easily find valuable resources and guidance on contributing to the cp-algorithms project.

The pages are compiled and published at [https://cp-algorithms.com](https://cp-algorithms.com).

## Steps to Contribute

Follow these steps to start contributing:

1. **Find the article you want to improve**. Click the pencil icon (:material-pencil:) next to the article title.
2. **Fork the repository** if prompted. This creates a copy of the repository in your GitHub account.
3. **Make your changes** directly in the GitHub editor or clone the repository to work locally.
4. **Preview your changes** using the [preview page](preview.md) to ensure they look correct.
5. **Commit your changes** by clicking the _Propose changes_ button.
6. **Create a Pull Request (PR)** by clicking _Compare & pull request_.
7. **Review process**: Someone from the core team will review your changes. This may take a few hours to a few days.

### Making Larger Changes

If you’re planning to make more significant changes, such as adding new articles or modifying multiple files:

- **Fork the project** using the traditional Git workflow (create a branch for your changes).
- **Edit files locally or in the GitHub UI**.
- **Submit a pull request** with your updates.

> For help with this workflow, check out this helpful guide: [Step-by-step guide to contributing on GitHub](https://www.dataschool.io/how-to-contribute-on-github/).

Author: I have changed the redirecting link from previous one to https://opensource.guide/how-to-contribute/.

### Updating Indexes

When you add new articles or reorganize existing ones, be sure to update the following files:

- **[navigation.md](https://github.com/cp-algorithms/cp-algorithms/blob/main/src/navigation.md)**: Update the list of all articles.

> - **[README.md](https://github.com/cp-algorithms/cp-algorithms/blob/main/README.md)**: Add new articles to the main project overview.

Author: The second line is better because it clearly instructs contributors to update the existing list of new articles, ensuring clarity in the required action.

- **Admonitions**: Use [admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#usage) for special content, such as theorems or examples.
- **Tabs**: Organize content with [content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#usage).
- **Tables**: Use [data tables](https://squidfunk.github.io/mkdocs-material/reference/data-tables/#usage) for organizing information.

Some advanced features may not be enabled or require a paid subscription. Keep this in mind when experimenting with formatting.

### Setting the HTML Title

By default, the first header (`# header`) of your article will be used as the HTML title. If your header contains a formula or complex text, you can manually set the title:

```markdown
---
title: Alternative HTML Title
---
# Proof of $a^2 + b^2 = c^2$
```

### Handling Redirects

If you move or rename an article, make sure to set up a redirect. A redirect file should look like this:

```md
<meta http-equiv="refresh" content="0; url=https://cp-algorithms.com/<new-section>/<new-article>.html">
# Article Name

> This article has been moved to a [new location](<new-section>/<new-article>.html).

Author: Using relative URLs with `[...]` enhances portability across different environments. For linking, it’s best to use `.html` for rendered pages and reserve `.md` for internal markdown references to ensure clarity and compatibility with MkDocs.

> This article has been moved to a [new location](<new-section>/<new-article>.html).

Reviewer: I think `.md` should make more sense because this is a markdown source file that is used by MkDocs to generate `.html` files, and the general convention for these templates seems to use `.md` in relative links. This also preserves correct links when viewed from GitHub markdown preview, although it's less important.

Also, you didn't change the redirect meta to the relative url...

### Redirects
### Setting the HTML Title

Files should not be moved or renamed without making redirects. A redirect page should generally look as follows:
By default, the first header (`# header`) of your article will be used as the HTML title. If your header contains a formula or complex text, you can manually set the title:

```md
<meta http-equiv="refresh" content="0; url=https://cp-algorithms.com/<new section>/<new name>.html">
# <Article name>
Article was moved (renamed). <a href="<new section>/<new name>.html">new URL</a>.
```markdown
---
title: Alternative HTML Title
---
# Proof of $a^2 + b^2 = c^2$
```

### Linking to a section with anchors
### Handling Redirects

Also it's kind of problematic when renaming a section of an article.
The section title is used for linking.
E.g. a section on the page `article.md` with
If you move or rename an article, make sure to set up a redirect. A redirect file should look like this:

```md
## Some title
<meta http-equiv="refresh" content="0; url=https://cp-algorithms.com/<new-section>/<new-article>.html">
# Article Name
This article has been moved to a [new location](<new-section>/<new-article>.html).
```

> can be linked to with `/article.html#some-title`.

Author: You're correct; links to sections can break with redirects. Standard HTML meta redirects do not preserve the fragment identifier (e.g., `#some-title`). To maintain links to specific sections after an article is moved, we would need a server-side redirect or JavaScript solution.

> We follow certain conventions across the project. For example, we agreed to use the `\binom{n}{k}` notation for binomial coefficients instead of `C_n^k` as outlined in [issue #83](https://github.com/cp-algorithms/cp-algorithms/issues/83).

Reviewer: Why did you resolve this? Nothing changed here...

> We follow certain conventions across the project. For example, we agreed to use the `\binom{n}{k}` notation for binomial coefficients instead of `C_n^k` as outlined in [issue #83](https://github.com/cp-algorithms/cp-algorithms/issues/83).

Reviewer: ```suggestion
We follow certain conventions across the project. For example, we agreed to use the `\binom{n}{k}` notation for binomial coefficients instead of `C_n^k` as outlined in [issue #83](https://github.com/cp-algorithms/cp-algorithms/issues/83). The first one renders as $\binom{n}{k}$ and is a more universal convention. The second would render as $C_n^k$.
```

>     $ git clone --recursive https://github.com/cp-algorithms/cp-algorithms.git && cd cp-algorithms

Author: Regarding making the $ non-selectable in MkDocs code blocks, it typically depends on the Markdown renderer being used.

2. Install dependencies and serve the site:

    ```console
    $ scripts/install-mkdocs.sh # requires pip
    $ mkdocs serve
    ```

   This will run the site locally so you can preview your changes. Note that some features are disabled in local builds.

### Optional Plugins

- **Git Revision Date Plugin**: Disabled by default but can be enabled with:

    ```console
    $ export MKDOCS_ENABLE_GIT_REVISION_DATE=True
    ```

- **Git Committers Plugin**: Disabled by default, as it requires a GitHub personal access token. Enable it like this:

    ```console
    $ export MKDOCS_ENABLE_GIT_COMMITTERS=True
    $ export MKDOCS_GIT_COMMITTERS_APIKEY=your_token_here
    ```

   You can generate your token [here](https://github.com/settings/tokens). Only public access permissions are needed.

## Testing Code Snippets

If your article includes code snippets, it’s helpful to include tests to ensure that they run correctly.

1. Name the code snippet:

    ```{.cpp file=snippet-name}
    // code here

>     ```

Reviewer: Why did you resolve this? It's still not fixed...

2. Run `extract_snippets.py` from the `test` directory to extract snippets into header files. Create a test file that includes these headers and checks their behavior.

3. You can run all tests with the `test.sh` script:

    ```console
    $ cd test
    $ ./test.sh

>     ```

Reviewer: I'd prefer to keep example output here.

```

This will allow existing links to continue working even after the section title is changed.

### Article Tags

We use tags to differentiate between original content and translated articles. Add the appropriate tag at the top of your article:

- **For original articles**:

    ```md
    ---
    tags:
        - Original
    ---
    ```

- **For translated articles**:

    ```md
    ---
    tags:
        - Translated
    e_maxx_link: <original-link>
    ---
    ```

    Replace `<original-link>` with the last part of the URL (e.g., for `http://e-maxx.ru/algo/euler_function`, use `euler_function`).

## Conventions

We follow certain conventions across the project. For example, we agreed to use the `\binom{n}{k}` notation for binomial coefficients instead of `C_n^k` as outlined in [issue #83](https://github.com/cp-algorithms/cp-algorithms/issues/83).

## Adding Problems

When adding problems, try to arrange them by difficulty. If you're unable to, don't worry—just add the problem, and someone else can adjust the order later.

## Local Development Setup

You can preview changes locally before pushing them to GitHub. To do this:

1. Clone the repository:

    ```console

>     git clone --recursive https://github.com/cp-algorithms/cp-algorithms.git && cd cp-algorithms

Reviewer: Please also remove prefix `$` in all other places in the article where applicable.

    ```

   This will run the site locally so you can preview your changes. Note that some features are disabled in local builds.

### Optional Plugins

- **Git Revision Date Plugin**: Disabled by default but can be enabled with:

    ```console
    $ export MKDOCS_ENABLE_GIT_REVISION_DATE=True
    ```

- **Git Committers Plugin**: Disabled by default, as it requires a GitHub personal access token. Enable it like this:

    ```console
    $ export MKDOCS_ENABLE_GIT_COMMITTERS=True
    $ export MKDOCS_GIT_COMMITTERS_APIKEY=your_token_here
    ```

   You can generate your token [here](https://github.com/settings/tokens). Only public access permissions are needed.

## Testing Code Snippets

If your article includes code snippets, it’s helpful to include tests to ensure that they run correctly.

1. Name the code snippet:
    ```{.cpp file=snippet-name}
    // code here
    ```
2. Run `extract_snippets.py` from the `test` directory to extract snippets into header files. Create a test file that includes these headers and checks their behavior.
3. You can run all tests with the `test.sh` script:
    ```console
    $ cd test
    $ ./test.sh
    ```
    **Example Output:**

>     ```
>     Running test_aho_corasick.cpp - Passed in 635 ms
>    Running test_balanced_brackets.cpp - Passed in 1390 ms
>     Running test_burnside_tori.cpp - Passed in 378 ms
>     ...
>     51 PASSED in 49.00 seconds
>     ```

Reviewer: ```suggestion
    ```
    Running test_aho_corasick.cpp - Passed in 635 ms
    Running test_balanced_brackets.cpp - Passed in 1390 ms
    Running test_burnside_tori.cpp - Passed in 378 ms
    ...
    51 PASSED in 49.00 seconds
    ```
```
Otherwise it misrenders.

- **Code blocks**: [Code blocks](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#usage) are great for adding code snippets in articles.
- **Admonitions**: Use [admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#usage) for special content, such as theorems or examples.
- **Tabs**: Organize content with [content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#usage).
- **Tables**: Use [data tables](https://squidfunk.github.io/mkdocs-material/reference/data-tables/#usage) for organizing information.

Some advanced features may not be enabled or require a paid subscription. Keep this in mind when experimenting with formatting.

### Setting the HTML Title

By default, the first header (`# header`) of your article will be used as the HTML title. If your header contains a formula or complex text, you can manually set the title:

```markdown
---
title: Alternative HTML Title
---
# Proof of $a^2 + b^2 = c^2$
```

### Handling Redirects

If you move or rename an article, make sure to set up a redirect. A redirect file should look like this:

```md

> <meta http-equiv="refresh" content="0; url=/<new-section>/<new-article>.html">
> # Article Name
> This article has been moved to a [new location](<new-section>/<new-article>.md).

Reviewer: ```suggestion
<meta http-equiv="refresh" content="0; url=../new-section/new-article.html">
# Article Name
This article has been moved to a [new location](new-section/new-article.md).
```

> <meta http-equiv="refresh" content="0; url=https://cp-algorithms.com/<new-section>/<new-article>.html">
> # Article Name
> This article has been moved to a [new location](<new-section>/<new-article>.html).

Reviewer: ```suggestion
<meta http-equiv="refresh" content="0; url=/<new-section>/<new-article>.html">
# Article Name
This article has been moved to a [new location](<new-section>/<new-article>.md).
```

```

### Maintaining Anchor Links

> If you rename a section, the link to that section might break. To avoid this, add an anchor manually:

Reviewer: ```suggestion
If you rename a section, the link to that section (`/article.html#section-title`) might break. To avoid this, add an anchor manually:
```

> If you rename a section, the link to that section (`/article.html#section-title`) might break. To avoid this, add an anchor manually:

Reviewer: ```suggestion
If you rename a section, the link to that section (`/article.html#old-section-title`) might break. To avoid this, add an anchor manually:
```

## Conventions

We follow certain conventions across the project. For example, we agreed to use the `\binom{n}{k}` notation for binomial coefficients instead of `C_n^k` as outlined in [issue #83](https://github.com/cp-algorithms/cp-algorithms/issues/83). The first one renders as $\binom{n}{k}$ and is a more universal convention. The second would render as $C_n^k$.

## Adding Problems

When adding problems, try to arrange them by difficulty. If you're unable to, don't worry—just add the problem, and someone else can adjust the order later.

## Local Development Setup

You can preview changes locally before pushing them to GitHub. To do this:

1. Clone the repository:

    ```console
    git clone --recursive https://github.com/cp-algorithms/cp-algorithms.git && cd cp-algorithms
    ```

2. Install dependencies and serve the site:

    ```console
    scripts/install-mkdocs.sh # requires pip
    mkdocs serve
    ```

   This will run the site locally so you can preview your changes. Note that some features are disabled in local builds.

### Optional Plugins

> - **Git Revision Date Plugin**: Disabled by default but can be enabled with:

Reviewer: ```suggestion
- **Git Revision Date Plugin**: Disabled by default, as it produces errors when you have uncommited changes in the working tree. Can be enabled with:
```

    ```console
    export MKDOCS_ENABLE_GIT_REVISION_DATE=True
    ```

- **Git Committers Plugin**: Disabled by default, as it requires a GitHub personal access token. Enable it like this:

    ```console
    export MKDOCS_ENABLE_GIT_COMMITTERS=True
    export MKDOCS_GIT_COMMITTERS_APIKEY=your_token_here
    ```

   You can generate your token [here](https://github.com/settings/tokens). Only public access permissions are needed.

## Testing Code Snippets

If your article includes code snippets, it’s helpful to include tests to ensure that they run correctly.

1. Name the code snippet:

> 
>    ```{.cpp file=snippet-name}
>     // code here
>     ```
>    

Reviewer: `````suggestion
````
```{.cpp file=snippet-name}
// code here
```
````
`````
